File: Wordle_CS/ConsoleApp/MainMenu.cs
using Wordle_CS.Players;

namespace Wordle_CS.ConsoleApp;

/// <summary>
/// Possible states of the main Wordle program.
/// </summary>
public enum ProgramState
{
    /// <summary>
    /// Request the user's login information
    /// </summary>
    LogIn,
    /// <summary>
    /// Run the main menu
    /// </summary>
    MainMenu,
    /// <summary>
    /// Delete the current user
    /// </summary>
    DeleteUser,
    /// <summary>
    /// Exit the program
    /// </summary>
    Exit,
}

/// <summary>
/// Possible selections in the main menu.
/// </summary>
enum UserSelection
{
    /// <summary>
    /// Play a game of Wordle
    /// </summary>
    PlayGame = 1,
    /// <summary>
    /// View the current player's statistics
    /// </summary>
    ViewStats,
    /// <summary>
    /// Log off
    /// </summary>
    LogOff,
    /// <summary>
    /// Delete the current user
    /// </summary>
    DeleteUser,
}

/// <summary>
/// Provides methods to manage the main menu of the Wordle
/// program, such as user login and running the game.
/// </summary>
public class MainMenu
{
    /// <summary>
    /// Requests a user to enter their login information.
    /// <para/>
    /// The user may choose to quit the program (or forcibly
    /// quit using Ctrl-C), in which case this function returns
    /// None. Otherwise, this function returns information
    /// about the player.
    /// <para/>
    /// If the user does not yet exist in the given databse,
    /// they will be added to it.
    /// </summary>
    /// <param name="usernames">A set of existing usernames</param>
    /// <returns>
    /// Information about the player, or <see langword="null"/> if the program
    /// should exit (either by user request or a database error)
    /// </returns>
    public static PlayerInfo? RequestUserLogin(SortedSet<string> usernames)
    {
        var username = RequestUsername(usernames);
        if (username is null)
        {
            return null;
        }

        PlayerInfo? playerInfo;
        try
        {
            playerInfo = PlayerInfo.CreateFromFile($"{username}.txt");
        }
        catch (IOException e)
        {
            Console.WriteLine(e.Message);
            return null;
        }

        Console.WriteLine($"Hello, {username}");

        var player = playerInfo ?? new PlayerInfo(username);
        return player;
    }

    /// <summary>
    /// Requests a user to enter their username.
    /// <para/>
    /// The user may choose to quit the program (or forcibly
    /// quit using Ctrl-C), in which case this function returns
    /// a nullopt.
    /// <para/>
    /// If the user does not yet exist in the given database,
    /// they will be added to it.
    /// </summary>
    /// <param name="usernames">A set of existing usernames</param>
    /// <returns>
    /// The player's username, or <see langword="null"/> if the program should
    /// exit (either by user request or a database error)
    /// </returns>
    private static string? RequestUsername(SortedSet<string> usernames)
    {
        if (usernames.Count != 0)
        {
            Console.WriteLine("List of existing users:");
            foreach (var name in usernames)
            {
                Console.WriteLine(name);
            }
            Console.WriteLine();
        }

        Console.WriteLine("Note: usernames are case-insensitive");
        Console.WriteLine("Type \":q\" to exit");
        Console.Write("Username: ");

        string? username = Console.ReadLine();
        if (username is null or ":q")
        {
            return null;
        }

        usernames.Add(username);

        return username;
    }

    /// <summary>
    /// Runs the Wordle main menu.
    /// <para/>
    /// The main menu gives the player four options:
    /// <list type="bullet">
    ///     <item>Play a game of Wordle</item>
    ///     <item>View their statistics</item>
    ///     <item>Log out</item>
    ///     <item>Delete their account</item>
    /// </list>
    /// <para/>
    /// This function lets the caller know what the next
    /// state of the program should be. For example, if
    /// the user has logged off, the main program should
    /// return to the login screen.
    /// </summary>
    /// <param name="currentPlayer">The logged-in player</param>
    /// <param name="dictionary">A dictionary of valid Wordle words</param>
    /// <returns></returns>
    public static ProgramState Run(PlayerInfo currentPlayer, HashSet<string> dictionary)
    {
        var userSelection = RequestUserSelection();
        if (userSelection is null)
        {
            return ProgramState.Exit;
        }

        switch (userSelection)
        {
            case UserSelection.PlayGame:
                var randWord = currentPlayer.GetRandomWord(dictionary);
                if (randWord is not null)
                {
                    WordleAnswer answer = new(randWord);
                    Game.Run(answer, currentPlayer, dictionary);
                    Console.WriteLine(currentPlayer.GetStats());
                    try
                    {
                        currentPlayer.WriteToFile($"{currentPlayer.Username}.txt");
                    }
                    catch (Exception)
                    {
                        Console.WriteLine("Error: could not write to user database file, progress has not been saved");
                    }
                }
                else
                {
                    /* couldn't get a word, player has already played every word */
                    Console.WriteLine("There are no remaining words in the dictionary.");
                }

                return ProgramState.MainMenu;

            case UserSelection.ViewStats:
                Console.WriteLine(currentPlayer.GetStats());
                return ProgramState.MainMenu;

            case UserSelection.LogOff:
                return ProgramState.LogIn;

            case UserSelection.DeleteUser:
                Console.Write($"Are you sure you would like to delete user: {currentPlayer.Username} [y/N] ");

                var userConfirmation = Console.ReadLine();
                if (userConfirmation?.ToLower()?.Trim() == "y")
                {
                    Console.WriteLine();
                    return ProgramState.DeleteUser;
                }
                else
                {
                    Console.WriteLine("Action aborted");
                    return ProgramState.MainMenu;
                }
        }

        return ProgramState.MainMenu;
    }

    /// <summary>
    /// Requests a user to input their selection.
    /// <para/>
    /// This function gives the player four options:
    /// <list type="bullet">
    ///     <item>Play a game of Wordle</item>
    ///     <item>View their statistics</item>
    ///     <item>Log out</item>
    ///     <item>Delete their account</item>
    /// </list>
    /// <para/>
    /// The user can terminate the program early using Ctrl-C,
    /// in which case this function returns a nullopt.
    /// </summary>
    /// <returns>The user selection, or <see langword="null"/> if none was given</returns>
    private static UserSelection? RequestUserSelection()
    {
        Console.WriteLine();
        Console.WriteLine("[1] Play a game of Wordle");
        Console.WriteLine("[2] View player statistics");
        Console.WriteLine("[3] Log off");
        Console.WriteLine("[4] Delete user");

        UserSelection? userSelection;
        while (true)
        {
            Console.Write("Selection: ");
            var selectionStr = Console.ReadLine();
            if (selectionStr is null)
            {
                userSelection = null;
                break;
            }

            if (int.TryParse(selectionStr, out int selection))
            {
                userSelection = (UserSelection)selection;
                break;
            }
            else
            {
                Console.WriteLine("Error: selection must be an integer");
            }
        }
        Console.WriteLine();

        return userSelection;
    }
}

File: wordle-rs/crates/wordle-terminal/src/main_menu.rs
//! This module provides methods to manage the main menu of
//! the Wordle program, such as user login and running the game.
//!
//! Author: Benjamin Hall

use std::{
    borrow::Borrow,
    collections::{BTreeSet, HashSet},
    io::{self, stdin, Write},
};

use crate::game;
use wordle::WordleAnswer;
use wordle_db::PlayerInfo;

/// Possible states of the main Wordle program.
#[derive(Default, Debug, Clone, PartialEq, Eq)]
pub enum ProgramState<S> {
    /// Request the user's login information
    #[default]
    LogIn,
    /// Run the main menu
    MainMenu(PlayerInfo<S>),
    /// Delete the current user
    DeleteUser(PlayerInfo<S>),
    /// Exit the program
    Exit,
}

/// Possible user selections in the main menu.
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
enum UserSelection {
    /// Play a game of Wordle
    PlayGame = 1,
    /// View the current player's statistics
    ViewStats,
    /// Log off
    LogOff,
    /// Delete the current user
    DeleteUser,
}

impl TryFrom<isize> for UserSelection {
    type Error = ();

    fn try_from(v: isize) -> Result<Self, Self::Error> {
        match v {
            x if x == Self::PlayGame as isize => Ok(Self::PlayGame),
            x if x == Self::ViewStats as isize => Ok(Self::ViewStats),
            x if x == Self::LogOff as isize => Ok(Self::LogOff),
            x if x == Self::DeleteUser as isize => Ok(Self::DeleteUser),
            _ => Err(()),
        }
    }
}

/// Requests a user to enter their login information.
///
/// The user may choose to quit the program (or forcibly
/// quit using Ctrl-C), in which case this function returns
/// None. Otherwise, this function returns information
/// about the player.
///
/// If the user does not yet exist in the given databse,
/// they will be added to it.
///
/// # Panics
///
/// Panics if writing to [`io::stdout`] fails.
///
/// # Examples
///
/// Basic usage:
/// ```
/// # use std::collections::BTreeSet;
/// # use wordle_terminal::main_menu;
/// # fn read_usernames(filename: &str) -> BTreeSet<String> {
/// # BTreeSet::new()
/// # }
/// let mut usernames: BTreeSet<String> =
///     read_usernames("usernames.txt");
///
/// let player_info = main_menu::request_user_login(&mut usernames);
/// match player_info {
///     Some(player) => { /* run program */ },
///     None => { /* exit program */ }
/// }
/// ```
#[must_use]
pub fn request_user_login(usernames: &mut BTreeSet<String>) -> Option<PlayerInfo<String>> {
    let Some(username) = request_username(usernames) else {
        /* user requested to exit the game */
        return None;
    };

    let player_info = match PlayerInfo::from_file(&(username.clone() + ".txt")) {
        Ok(player_info) => player_info,
        /* error reading the database file */
        Err(msg) => {
            println!("{msg}");
            return None;
        }
    };

    println!("Hello, {username}");

    /* this might be a new user, create a fresh instance of PlayerInfo if so */
    player_info.or_else(|| Some(PlayerInfo::new(username)))
}

/// Requests a user to enter their username.
///
/// The user may choose to quit the program (or forcibly
/// quit using Ctrl-C), in which case this function returns
/// None.
///
/// If the user does not yet exist in the given database,
/// they will be added to it.
///
/// # Panics
///
/// Panics if writing to [`io::stdout`] fails.
///
/// # Examples
///
/// Basic usage:
/// ```ignore
/// # use std::collections::BTreeSet;
/// # use wordle_terminal::main_menu;
/// # fn read_usernames(filename: &str) -> BTreeSet<String> {
/// # BTreeSet::new()
/// # }
/// let mut usernames: BTreeSet<String> =
///     read_usernames("usernames.txt");
///
/// let username = main_menu::request_username(&mut usernames);
/// match username {
///     Some(username) => { /* create player with username */ },
///     None => { /* exit program */ }
/// }
/// ```
#[must_use]
fn request_username(usernames: &mut BTreeSet<String>) -> Option<String> {
    {
        let mut lock = io::stdout().lock();
        if !usernames.is_empty() {
            writeln!(lock, "List of existing users:").unwrap();
            for name in usernames.iter() {
                writeln!(lock, "{name}").unwrap();
            }
            writeln!(lock).unwrap();
        }

        writeln!(lock, "Note: usernames are case-insensitive").unwrap();
        writeln!(lock, "Type \":q\" to exit").unwrap();
        write!(lock, "Username: ").unwrap();
        lock.flush().unwrap();
    }

    let mut username = String::new();
    if stdin().read_line(&mut username).is_err() {
        /* user likely quit the program with Ctrl-C */
        return None;
    }
    let username = username.trim().to_lowercase();

    if username == ":q" {
        /* user wants to exit */
        return None;
    }

    if !usernames.contains(&username) {
        /* new user, add to database */
        usernames.insert(username.clone());
    }

    Some(username)
}

/// Runs the Wordle main menu.
///
/// The main menu gives the player four options:
/// - Play a game of Wordle
/// - View their statistics
/// - Log out
/// - Delete their account
///
/// This function lets the caller know what the next
/// state of the program should be. For example, if
/// the user has logged off, the main program should
/// return to the login screen.
///
/// # Panics
///
/// Panics if writing to [`io::stdout`] fails.
///
/// # Examples
///
/// Basic usage:
/// ```no_run
/// # use std::collections::{BTreeSet, HashSet};
/// # use wordle_terminal::main_menu;
/// # fn read_dictionary(filename: &str) -> HashSet<String> {
/// # HashSet::new()
/// # }
/// # fn read_usernames(filename: &str) -> BTreeSet<String> {
/// # BTreeSet::new()
/// # }
/// let dictionary: HashSet<String> =
///     read_dictionary("dictionary.txt");
/// let mut usernames: BTreeSet<String> =
///     read_usernames("usernames.txt");
///
/// let player_info = main_menu::request_user_login(&mut usernames);
/// if let Some(player_info) = player_info {
///     let next_state = main_menu::run(player_info, &dictionary);
/// }
/// ```
#[must_use]
pub fn run<S>(
    mut player: PlayerInfo<S>,
    dictionary: &HashSet<String, impl std::hash::BuildHasher>,
) -> ProgramState<S>
where
    S: Borrow<str>,
{
    let Some(user_selection) = request_user_selection() else {
        /* user likely quit the program with Ctrl-C */
        return ProgramState::Exit;
    };

    match user_selection {
        UserSelection::PlayGame => {
            /* run a game of Wordle */
            if let Some(answer) = player.get_random_word(dictionary) {
                let answer = WordleAnswer::new(answer);
                game::run(&answer, &mut player, dictionary);
                /* print the player's statistics after the game ends */
                println!("{}", player.get_stats());
                /* save the user's new statistics to their database */
                if player
                    .write_to_file(&(player.get_username().to_owned() + ".txt"))
                    .is_err()
                {
                    /* report that we could not write to the database, but do not exit */
                    println!(
                        "Error: could not write to user database file, progress has not been saved"
                    );
                }
            } else {
                /* couldn't get a word, player has already played every word */
                println!("There are no remaining words in the dictionary.");
            }
            ProgramState::MainMenu(player)
        }
        UserSelection::ViewStats => {
            println!("{}", player.get_stats());
            ProgramState::MainMenu(player)
        }
        /* user is logged off, go back to login screen */
        UserSelection::LogOff => ProgramState::LogIn,
        UserSelection::DeleteUser => {
            print!(
                "Are you sure you would like to delete user: {} [y/N] ",
                player.get_username()
            );
            io::stdout().flush().unwrap();

            let mut user_confirmation = String::new();
            match stdin().read_line(&mut user_confirmation) {
                Ok(_)
                    if {
                        user_confirmation.make_ascii_lowercase();
                        user_confirmation.trim()
                    } == "y" =>
                {
                    println!();
                    ProgramState::DeleteUser(player)
                }
                _ => {
                    println!("Action aborted");
                    ProgramState::MainMenu(player)
                }
            }
        }
    }
}

/// Requests a user to input their selection.
///
/// This function gives the player four options:
/// - Play a game of Wordle
/// - View their statistics
/// - Log out
/// - Delete their account
///
/// The user can terminate the program early using Ctrl-C,
/// in which case this function returns None.
///
/// # Panics
///
/// Panics if writing to [`io::stdout`] fails.
///
/// # Examples
///
/// Basic usage:
/// ```ignore
/// # use wordle_terminal::main_menu;
/// let user_selection = main_menu::request_user_selection();
/// match (user_selection) {
///     Some(user_selection) => { /* run selection */ },
///     None => { /* exit program */ }
/// }
/// ```
#[must_use]
fn request_user_selection() -> Option<UserSelection> {
    {
        let mut lock = io::stdout().lock();
        writeln!(lock).unwrap();
        writeln!(lock, "[1] Play a game of Wordle").unwrap();
        writeln!(lock, "[2] View player statistics").unwrap();
        writeln!(lock, "[3] Log off").unwrap();
        writeln!(lock, "[4] Delete user").unwrap();
    }

    let user_selection = loop {
        print!("Selection: ");
        io::stdout().flush().unwrap();

        let mut selection_str = String::new();
        if stdin().read_line(&mut selection_str).is_err() {
            /* user likely quit the program with Ctrl-C */
            break None;
        }

        if let Ok(selection) = selection_str.trim().parse::<isize>() {
            let user_selection = UserSelection::try_from(selection).ok();
            if user_selection.is_none() {
                /* selection out of range */
                println!("Error: invalid selection");
            } else {
                /* valid selection, stop the read loop */
                break user_selection;
            }
        } else {
            println!("Error: selection must be an integer");
        }
    };
    println!();

    user_selection
}
